core/deprecation_manager.cpp

```diff
@@ -0,0 +1,90 @@
+#include "deprecation_manager.h"
+#include "core/class_db.h"
+
+#ifdef TOOLS_ENABLED
+
+DeprecationInfo::DeprecationInfo() :
+		warned(false),
+		custom_warning("") {
+}
+
+DeprecationInfo::DeprecationInfo(String p_custom_warning) :
+		warned(false),
+		custom_warning(p_custom_warning) {
+}
+
+HashMap<String, DeprecationManager::ClassInfo> DeprecationManager::classes;
+
+void DeprecationManager::warn_deprecated(const StringName &p_name, const StringName &p_type, const DeprecationInfo &p_info) {
+	String warning;
+
+	if (!p_info.custom_warning.empty()) {
+		warning = p_info.custom_warning;
```

Comment:
Just an opinion here. I think it would better for the custom warning to be something to append to the default warning. For example, if I have a custom warning _"Use Bar instead"_, it should print _"The method Foo has been deprecated and will be removed in the future. Use Bar instead."_.

Reply:
That works as well, I was just trying to provide the  most versatility that maybe one would want to print a totally different message.

Reply:
I changed it to what you suggested.